File: multilibrary/src/main/java/com/zhongjh/albumcamerarecorder/camera/ui/camera/state/type/VideoMultipleIn.java
package com.zhongjh.albumcamerarecorder.camera.ui.camera.state.type;

import com.zhongjh.albumcamerarecorder.camera.ui.camera.BaseCameraFragment;
import com.zhongjh.albumcamerarecorder.camera.ui.camera.manager.CameraVideoManager;
import com.zhongjh.albumcamerarecorder.camera.ui.camera.state.CameraStateManager;
import com.zhongjh.albumcamerarecorder.camera.ui.camera.state.StateMode;
import com.zhongjh.albumcamerarecorder.camera.ui.camera.manager.CameraPictureManager;

/**
 * 多个视频的状态录制中
 *
 * @author zhongjh
 * @date 2021/11/29
 */
public class VideoMultipleIn extends StateMode {

    /**
     * @param cameraFragment     主要是多个状态围绕着cameraLayout进行相关处理
     * @param cameraStateManager 可以让状态更改别的状态
     */
    public VideoMultipleIn(BaseCameraFragment<? extends CameraStateManager,
            ? extends CameraPictureManager,
            ? extends CameraVideoManager> cameraFragment, CameraStateManager cameraStateManager) {
        super(cameraFragment, cameraStateManager);
    }

    @Override
    public String getName() {
        return "VideoMultipleIn";
    }

    @Override
    public void onActivityPause() {
        getCameraFragment().getCameraVideoManager().getVideoTimes().clear();
        // 重置所有
        getCameraFragment().resetStateAll();
        // 恢复预览状态
        getCameraStateManagement().setState(getCameraStateManagement().getPreview());
    }

    @Override
    public Boolean onBackPressed() {
        getCameraFragment().getCameraManage().closeVideo();
        getCameraFragment().getPhotoVideoLayout().resetConfirm();
        getCameraFragment().getPhotoVideoLayout().getViewHolder().btnClickOrLong.selectionRecordRollBack();

        if (getCameraFragment().getCameraVideoManager().getVideoTimes().isEmpty()) {
            // 如果没有视频节点则重置所有按钮
            getCameraFragment().getPhotoVideoLayout().reset();
            // 恢复预览状态
            getCameraStateManagement().setState(getCameraStateManagement().getPreview());
        } else {
            // 如果有视频节点则中断中心按钮
            getCameraFragment().getPhotoVideoLayout().getViewHolder().btnClickOrLong.breakOff();
            // 恢复预览状态
            getCameraStateManagement().setState(getCameraStateManagement().getVideoMultiple());
        }
        return true;
    }

    @Override
    public boolean onActivityResult(int resultCode) {
        return true;
    }

    @Override
    public void pauseRecord() {
        // 切回非录制中的状态
        getCameraStateManagement().setState(getCameraStateManagement().getVideoMultiple());
        getCameraFragment().getCameraManage().pauseVideo();
    }

    @Override
    public void onLongClickFinish() {
        getCameraFragment().getCameraManage().stopVideo();
    }

}

File: multilibrary/src/main/java/com/zhongjh/multimedia/camera/ui/camera/state/type/VideoMultipleIn.kt
package com.zhongjh.multimedia.camera.ui.camera.state.type

import com.zhongjh.multimedia.camera.ui.camera.BaseCameraFragment
import com.zhongjh.multimedia.camera.ui.camera.manager.CameraPictureManager
import com.zhongjh.multimedia.camera.ui.camera.manager.CameraVideoManager
import com.zhongjh.multimedia.camera.ui.camera.state.CameraStateManager
import com.zhongjh.multimedia.camera.ui.camera.state.type.impl.StateMode

/**
 * 视频录制中
 *
 * @param cameraFragment     主要是多个状态围绕着cameraLayout进行相关处理
 * @param cameraStateManager 可以让状态更改别的状态
 *
 * @author zhongjh
 * @date 2021/11/29
 */
class VideoMultipleIn(cameraFragment: BaseCameraFragment<out CameraStateManager, out CameraPictureManager, out CameraVideoManager>, cameraStateManager: CameraStateManager) :
    StateMode(cameraFragment, cameraStateManager) {
    override fun getName(): String {
        return "VideoMultipleIn"
    }

    override fun onActivityPause() {
        cameraFragment.cameraVideoManager.videoTime = 0L
        // 重置所有
        cameraFragment.resetStateAll()
        // 恢复预览状态
        stateManagerRef.get()?.let { stateManager ->
            stateManager.state = stateManager.preview
        }
    }

    override fun onBackPressed(): Boolean {
        cameraFragment.cameraManage.closeVideo()
        cameraFragment.photoVideoLayout.resetConfirm()

        if (cameraFragment.cameraVideoManager.videoTime == 0L) {
            // 如果没有视频节点则重置所有按钮
            cameraFragment.photoVideoLayout.reset()
            // 恢复预览状态
            stateManagerRef.get()?.let { stateManager ->
                stateManager.state = stateManager.preview
            }
        } else {
            // 如果有视频节点则中断中心按钮
            cameraFragment.photoVideoLayout.viewHolder.btnClickOrLong.breakOff()
            // 恢复录制状态
            stateManagerRef.get()?.let { stateManager ->
                stateManager.state = stateManager.videoMultiple
            }
        }
        return true
    }

    override fun pauseRecord() {
        // 切回非录制中的状态
        stateManagerRef.get()?.let { stateManager ->
            stateManager.state = stateManager.videoMultiple
        }
        cameraFragment.cameraManage.pauseVideo()
    }

    override fun onLongClickFinish() {
        cameraFragment.cameraManage.stopVideo()
    }
}
